### PR TITLE
Ubuntu 20.04: Fix systemvm cannot start up

### PR DESCRIPTION
**Pull request by ustcweizhou**
## Description

When use Ubuntu 20.04 as hypervisor, and create an advanced zone with security groups, the systemvms (ssvm and cpvm) are started but the cloud service is not be Up.

in /var/log/cloud.log, it says vm is unable to connect to management server on port 8250.

Later I found there is caused by a iptables rule below
```
-A BF-cloudbr0 -m physdev --physdev-out ens3: --physdev-is-bridged -j ACCEPT
```

The physical device should be "ens3" and "ens3:", which comes from
```
root@node62:~# bridge -o link show | awk '/master cloudbr0 / && !/^[0-9]+: vnet/ {print $2}' | head -1
ens3:
```
There is no issue with ubuntu 16.04
```
root@node12:~# bridge -o link show | awk '/master cloudbr0 / && !/^[0-9]+: vnet/ {print $2}' | head -1
eth0
```

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

### CODE REVIEW

File: scripts/vm/network/security_group.py [hunk 1]
```diff
@@ -510,7 +510,7 @@ def check_default_network_rules(vm_name, vm_id, vm_ip, vm_ip6, vm_mac, vif, brna
         rules = execute("iptables-save |grep -w %s |grep -w %s |grep -w %s" % (brfw, vif, vmchain_default))
     except:
         rules = None
-    if rules is None or rules is "":
+    if rules is None or rules == "":
```
Comment (wido) [inline scripts/vm/network/security_group.py:513]: Maybe better to change this to:

<pre>if not rules:</pre>

That captures the None and empty string case right away.

File: scripts/vm/network/security_group.py [hunk 2]
```diff
@@ -185,7 +185,7 @@ def destroy_network_rules_for_nic(vm_name, vm_ip, vm_mac, vif, sec_ips):
         logging.debug("Ignoring failure to delete ebtable rules for vm: " + vm_name)
 
 def get_bridge_physdev(brname):
-    physdev = execute("bridge -o link show | awk '/master %s / && !/^[0-9]+: vnet/ {print $2}' | head -1" % brname)
+    physdev = execute("bridge -o link show | awk '/master %s / && !/^[0-9]+: vnet/ {print $2}' | head -1 | cut -d ':' -f1" % brname)
```
Comment (wido) [inline scripts/vm/network/security_group.py:188]: Do we know what changed in Ubuntu 20.04? As the '-o' flag should make it stable to use on the CLI

File: scripts/vm/network/security_group.py (hunk 1 above)
Comment (weizhouapache) [reply scripts/vm/network/security_group.py:513]: good, thanks @wido

**Comment by wido** [comment]
LGTM

File: scripts/vm/network/security_group.py (hunk 2 above)
Comment (rhtyd) [reply scripts/vm/network/security_group.py:188]: Should we use `ip` (iproute2) instead of old tools?

Comment (wido) [reply scripts/vm/network/security_group.py:188]: bridge is a part of iproute2. The old tool is called 'brctl'

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
cc @davidjumani can you review?

**Comment by davidjumani** [comment]
Changes LGTM. Will test!

**Comment by davidjumani** [comment]
Systemvm is still unable to come up on u20, but i think that it might be unrelated
```
2020-09-16 05:00:45,295 INFO  [kvm.storage.LibvirtStorageAdaptor] (agentRequest-Handler-5:null) (logid:287cef54) Trying to fetch storage pool 0e678d73-abb2-3295-b6c3-b6949a8a2f8d from libvirt
2020-09-16 05:00:48,388 INFO  [kvm.resource.LibvirtConnection] (agentRequest-Handler-1:null) (logid:287cef54) No existing libvirtd connection found. Opening a new one
2020-09-16 05:00:48,389 WARN  [kvm.resource.LibvirtConnection] (agentRequest-Handler-1:null) (logid:287cef54) Can not find a connection for Instance s-61-VM. Assuming the default connection.
2020-09-16 05:00:48,461 WARN  [kvm.resource.LibvirtKvmAgentHook] (agentRequest-Handler-1:null) (logid:287cef54) Groovy script '/etc/cloudstack/agent/hooks/libvirt-vm-state-change.groovy' is not available. Transformations will not be applied.
2020-09-16 05:00:48,461 WARN  [kvm.resource.LibvirtKvmAgentHook] (agentRequest-Handler-1:null) (logid:287cef54) Groovy scripting engine is not initialized. Data transformation skipped.
2020-09-16 05:01:18,900 INFO  [kvm.storage.LibvirtStorageAdaptor] (agentRequest-Handler-2:null) (logid:3d0925b0) Trying to fetch storage pool 0e678d73-abb2-3295-b6c3-b6949a8a2f8d from libvirt
2020-09-16 05:01:18,955 INFO  [kvm.storage.LibvirtStorageAdaptor] (agentRequest-Handler-2:null) (logid:3d0925b0) Attempting to remove volume a2fcc05f-2abd-44a5-a3cf-2474232dac1f from pool 0e678d73-abb2-3295-b6c3-b6949a8a2f8d
2020-09-16 05:01:43,249 INFO  [kvm.storage.LibvirtStorageAdaptor] (agentRequest-Handler-4:null) (logid:287cef54) Trying to fetch storage pool fb4bb1df-cb02-3dfe-9a79-d37ee241d200 from libvirt
2020-09-16 05:01:43,252 INFO  [kvm.storage.LibvirtStorageAdaptor] (agentRequest-Handler-4:null) (logid:287cef54) Trying to fetch storage pool fb4bb1df-cb02-3dfe-9a79-d37ee241d200 from libvirt
2020-09-16 05:01:43,264 INFO  [kvm.storage.LibvirtStorageAdaptor] (agentRequest-Handler-4:null) (logid:287cef54) Creating volume 96de7039-b33f-4dc3-9381-dc62f19dc4f2 from template 01e1c91c-f125-42b7-8e95-3d5145c0d0b3 in pool fb4bb1df-cb02-3dfe-9a79-d37ee241d200 (NetworkFilesystem) with size (0 bytes) 0
2020-09-16 05:01:43,265 INFO  [kvm.storage.LibvirtStorageAdaptor] (agentRequest-Handler-4:null) (logid:287cef54) Attempting to create volume 96de7039-b33f-4dc3-9381-dc62f19dc4f2 (NetworkFilesystem) in pool fb4bb1df-cb02-3dfe-9a79-d37ee241d200 with size (2.44 GB) 2621440000
2020-09-16 05:01:43,778 INFO  [kvm.storage.LibvirtStorageAdaptor] (agentRequest-Handler-5:null) (logid:287cef54) Trying to fetch storage pool fb4bb1df-cb02-3dfe-9a79-d37ee241d200 from libvirt
2020-09-16 05:01:43,814 INFO  [kvm.storage.LibvirtStorageAdaptor] (agentRequest-Handler-5:null) (logid:287cef54) Trying to fetch storage pool fb4bb1df-cb02-3dfe-9a79-d37ee241d200 from libvirt
2020-09-16 05:01:43,859 WARN  [kvm.resource.LibvirtKvmAgentHook] (agentRequest-Handler-5:null) (logid:287cef54) Groovy script '/etc/cloudstack/agent/hooks/libvirt-vm-xml-transformer.groovy' is not available. Transformations will not be applied.
2020-09-16 05:01:43,859 WARN  [kvm.resource.LibvirtKvmAgentHook] (agentRequest-Handler-5:null) (logid:287cef54) Groovy scripting engine is not initialized. Data transformation skipped.
2020-09-16 05:01:44,655 WARN  [resource.wrapper.LibvirtStartCommandWrapper] (agentRequest-Handler-5:null) (logid:287cef54) LibvirtException 
org.libvirt.LibvirtException: internal error: qemu unexpectedly closed the monitor: 2020-09-16T05:01:44.138805Z qemu-system-x86_64: error: failed to set MSR 0xe1 to 0x0
qemu-system-x86_64: /build/qemu-J4cYl4/qemu-4.2/target/i386/kvm.c:2691: kvm_buf_set_msrs: Assertion `ret == cpu->kvm_msr_buf->nmsrs' failed.
	at org.libvirt.ErrorHandler.processError(Unknown Source)
	at org.libvirt.Connect.processError(Unknown Source)
	at org.libvirt.Connect.processError(Unknown Source)
	at org.libvirt.Connect.domainCreateXML(Unknown Source)
	at com.cloud.hypervisor.kvm.resource.LibvirtComputingResource.startVM(LibvirtComputingResource.java:1612)
	at com.cloud.hypervisor.kvm.resource.wrapper.LibvirtStartCommandWrapper.execute(LibvirtStartCommandWrapper.java:85)
	at com.cloud.hypervisor.kvm.resource.wrapper.LibvirtStartCommandWrapper.execute(LibvirtStartCommandWrapper.java:45)
	at com.cloud.hypervisor.kvm.resource.wrapper.LibvirtRequestWrapper.execute(LibvirtRequestWrapper.java:78)
	at com.cloud.hypervisor.kvm.resource.LibvirtComputingResource.executeRequest(LibvirtComputingResource.java:1644)
	at com.cloud.agent.Agent.processRequest(Agent.java:645)
	at com.cloud.agent.Agent$AgentRequestHandler.doTask(Agent.java:1063)
	at com.cloud.utils.nio.Task.call(Task.java:83)
	at com.cloud.utils.nio.Task.call(Task.java:29)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
2020-09-16 05:01:44,779 INFO  [kvm.storage.LibvirtStorageAdaptor] (agentRequest-Handler-5:null) (logid:287cef54) Trying to fetch storage pool fb4bb1df-cb02-3dfe-9a79-d37ee241d200 from libvirt
2020-09-16 05:01:48,900 INFO  [kvm.resource.LibvirtConnection] (agentRequest-Handler-1:null) (logid:287cef54) No existing libvirtd connection found. Opening a new one
2020-09-16 05:01:48,901 WARN  [kvm.resource.LibvirtConnection] (agentRequest-Handler-1:null) (logid:287cef54) Can not find a connection for Instance s-62-VM. Assuming the default connection.
2020-09-16 05:01:48,970 WARN  [kvm.resource.LibvirtKvmAgentHook] (agentRequest-Handler-1:null) (logid:287cef54) Groovy script '/etc/cloudstack/agent/hooks/libvirt-vm-state-change.groovy' is not available. Transformations will not be applied.
2020-09-16 05:01:48,971 WARN  [kvm.resource.LibvirtKvmAgentHook] (agentRequest-Handler-1:null) (logid:287cef54) Groovy scripting engine is not initialized. Data transformation skipped.
2020-09-16 05:02:13,275 INFO  [kvm.storage.LibvirtStorageAdaptor] (agentRequest-Handler-2:null) (logid:57bb906e) Trying to fetch storage pool 0e678d73-abb2-3295-b6c3-b6949a8a2f8d from libvirt
2020-09-16 05:02:13,299 INFO  [kvm.storage.LibvirtStorageAdaptor] (agentRequest-Handler-2:null) (logid:57bb906e) Trying to fetch storage pool 0e678d73-abb2-3295-b6c3-b6949a8a2f8d from libvirt
2020-09-16 05:02:13,344 WARN  [kvm.resource.LibvirtKvmAgentHook] (agentRequest-Handler-2:null) (logid:57bb906e) Groovy script '/etc/cloudstack/agent/hooks/libvirt-vm-xml-transformer.groovy' is not available. Transformations will not be applied.
2020-09-16 05:02:13,344 WARN  [kvm.resource.LibvirtKvmAgentHook] (agentRequest-Handler-2:null) (logid:57bb906e) Groovy scripting engine is not initialized. Data transformation skipped.
2020-09-16 05:02:14,100 WARN  [resource.wrapper.LibvirtStartCommandWrapper] (agentRequest-Handler-2:null) (logid:57bb906e) LibvirtException 
org.libvirt.LibvirtException: internal error: qemu unexpectedly closed the monitor: 2020-09-16T05:02:13.656588Z qemu-system-x86_64: error: failed to set MSR 0xe1 to 0x0
```

**Comment by rhtyd** [comment]
@ustcweizhou can you test in your env and provide results?
@davidjumani can you test manually and see if it works/doesn't work?

**Comment by rhtyd** [comment]
ping @weizhouapache @ustcweizhou @davidjumani @shwstppr

**Comment by weizhouapache** [comment]
> qemu-system-x86_64: error: failed to set MSR 0xe1 to 0x0

@davidjumani 
I do not have this issue on my testing env.

are you using nested virtualization on a host with AMD cpu ?
you can try this workaround
```
echo 1 > /sys/modules/kvm/parameters/ignore_msrs
```
see https://patchwork.kernel.org/patch/42605/ for more details

**Comment by rhtyd** [comment]
ping @davidjumani can you try again?

**Review by davidjumani — APPROVED** [review]
LGTM! Systemvms come up

**Comment by rhtyd** [comment]
Merged based on Wido and David 's lgtms

File: scripts/vm/network/security_group.py (hunk 2 above)
Comment (abdelouahabb) [reply scripts/vm/network/security_group.py:188]: there is the trick to use `ls /sys/class/net` to get the exact interface names

**Pull request merged**